ironfish-cli/src/commands/service/snapshot.ts

```diff
@@ -0,0 +1,201 @@
+import path from 'path'
+import tar from 'tar'
+import { IronfishCommand } from '../../command'
+import { RemoteFlags } from '../../flags'
+import { ProgressBar } from '../../types'
+
+export default class CreateSnapshot extends IronfishCommand {
+  static hidden = true
+
+  static description = `Upload chain snapshot to a public bucket`
+
+  static flags = {
+    ...RemoteFlags,
+    bucket: Flags.string({
+      char: 'b',
+      parse: (input: string) => Promise.resolve(input.trim()),
+      required: false,
+      description: 'S3 bucket to upload snapshot to',
+    }),
+    path: Flags.string({
+      char: 'p',
+      parse: (input: string): Promise<string> => Promise.resolve(input.trim()),
+      required: false,
+      description: 'The path where the snapshot should be saved',
+    }),
+    maxBlocksPerChunk: Flags.integer({
+      char: 'm',
+      required: false,
+      default: isNaN(Number(process.env.MAX_BLOCKS_PER_SNAPSHOT_CHUNK))
+        ? 1000
+        : Number(process.env.MAX_BLOCKS_PER_SNAPSHOT_CHUNK),
+      description: 'The max number of blocks per file in the zipped snapshot',
+    }),
+  }
+
+  async start(): Promise<void> {
+    const { flags } = await this.parse(CreateSnapshot)
+
+    const bucket = (flags.bucket || process.env.IRONFISH_SNAPSHOT_BUCKET || '').trim()
+
+    let exportDir
+
+    if (flags.path) {
+      exportDir = this.sdk.fileSystem.resolve(flags.path)
+    } else {
+      try {
+        exportDir = await fsAsync.mkdtemp(`${os.tmpdir()}${path.sep}`)
+      } catch (err) {
+        this.log(`Could not create temp folder for snapshot generation`)
+        this.exit(1)
+      }
+    }
+    Assert.isNotUndefined(exportDir)
+
+    const blockExportPath = this.sdk.fileSystem.join(exportDir, 'blocks')
+    await this.sdk.fileSystem.mkdir(blockExportPath, { recursive: true })
+
+    this.log('Connecting to node...')
+
+    // TODO: There's a significant slowdown in the export process when running a
+    // full node. This may be due to CPU starvation or message formatting calls
+    // due to the use of node-ipc. We should revisit this in the future to allow
+    // for export without shutting down the node. -- deekerno
+    const client = await this.sdk.connectRpc(true)
+
+    const response = client.snapshotChainStream({
+      maxBlocksPerChunk: flags.maxBlocksPerChunk,
+    })
+
+    const { start, stop } = await AsyncUtils.first(response.contentStream())
+    this.log(`Retrieving blocks from ${start} -> ${stop} for snapshot generation`)
+
+    const progress = CliUx.ux.progress({
+      format: 'Retrieving blocks: [{bar}] {value}/{total} {percentage}% | ETA: {eta}s',
+    }) as ProgressBar
+
+    progress.start(stop - start + 1, 0)
+
+    for await (const result of response.contentStream()) {
+      if (result.buffer && result.seq) {
+        const blockFilePath = this.sdk.fileSystem.join(blockExportPath, `${result.seq}`)
+        await fsAsync.writeFile(blockFilePath, Buffer.from(result.buffer))
```

Comment:
In both cases, I can generate a UUID and use that to ensure that the block folders don't collide; does that sound good?

Reply:
That answers the first question. My second question around cleaning up the `blocks` directory is more about disk space usage -- the export process dumps out 10s-GB of intermediate block files, then zips them up into another 10s-GB archive. It makes sense to keep the archive around, since users will want to do something with it, but are the intermediate block files still needed?

Reply:
No, those files are necessarily needed after the fact as they're contained in the archive. I'll add functionality to remove them.